src/Microsoft.Diagnostics.EventFlow.Core/Implementations/LossReportingPropagatorBlock.cs

```diff
@@ -0,0 +1,209 @@
+            }
+
+            // We do not support re-linking
+            return new EmptyDisposable();
+        }
+
+        public DataflowMessageStatus OfferMessage(DataflowMessageHeader messageHeader, T messageValue, ISourceBlock<T> source, bool consumeToAccept)
+        {
+            VerifySourceInitiatedOperation(source);
+
+            if (this.completed)
+            {
+                return DataflowMessageStatus.DecliningPermanently;
+            }
+
+            var targetResponse = this.target.OfferMessage(messageHeader, messageValue, this, consumeToAccept);
+            switch(targetResponse)
+            {
+                case DataflowMessageStatus.Accepted:
+                    // Great, nothing to do
+                    break;
+
+                case DataflowMessageStatus.Declined:
+                case DataflowMessageStatus.DecliningPermanently:
+                case DataflowMessageStatus.NotAvailable:
+                    this.healthReporter.ReportThrottling();
+                    break;
+
+                case DataflowMessageStatus.Postponed:
+                    // This is a bit tricky--the message might be successfully consumed later, or it might no longer be available.
+                    // We'll know for sure based on the success/failure of ConsumeMessage().
+                    break;
+            }
+
+            return targetResponse;
+        }
+
+        public T ConsumeMessage(DataflowMessageHeader messageHeader, ITargetBlock<T> target, out bool messageConsumed)
+        {
+            VerifyTargetInitiatedOperation(target);
+
+            messageConsumed = false;
+
+            if (this.completed)
+            {
+                return default(T);
+            }
+
+            T message = this.source.ConsumeMessage(messageHeader, this, out bool consumedByMe);
+            if (consumedByMe)
+            {
+                messageConsumed = true;
+                return message;
+            }
+            else
+            {
+                // Message is no longer available
+                this.healthReporter.ReportThrottling();
+                return default(T);
+            }
+        }
+
+        public void ReleaseReservation(DataflowMessageHeader messageHeader, ITargetBlock<T> target)
+        {
+            VerifyTargetInitiatedOperation(target);
+
+            if (!this.completed)
+            {
+                this.source.ReleaseReservation(messageHeader, this);
+            }
+        }
+
+        public bool ReserveMessage(DataflowMessageHeader messageHeader, ITargetBlock<T> target)
+        {
+            VerifyTargetInitiatedOperation(target);
+
+            if (this.completed)
+            {
+                return false;
+            }
+
+            bool reserved = this.source.ReserveMessage(messageHeader, this);
+            if (!reserved)
+            {
+                // If we cannot reserve the message, it means it is gone before our target had a chance to consume it.
```

Comment:
This is not true. Source can return false if the message is already reserved by some other targets.